### PR TITLE
[BUGFIX release] Deprecate `@each` as a leaf node.

### PR DESCRIPTION
**Pull request by rwjblue**
Update stable version with a deprecation when using `foo.@each` as a dependent key. This code is removed on [beta](https://github.com/emberjs/ember.js/pull/11990) (2.0.0) and [canary](https://github.com/emberjs/ember.js/pull/11984) (2.1.0) branches already.

### CODE REVIEW

**Comment by rauhryan** [comment]
@rwjblue just out of curiousity. 

Is it still possible to track a sub-property of things in the array that are one level deep. 

I assume what used to be

`Ember.computed('foos.@each.bar')`

is changed to 

`Ember.computed('foos.[].bar')`

and all is well?

**Comment by stefanpenner** [comment]
@rauhryan other way around:

`Ember.computed('foos.[].bar')` becomes `Ember.computed('foos.@each.bar')
`Ember.computed('foos.@each')` becomes `Ember.computed('foos.[]')

**Comment by rauhryan** [comment]
thanks for clarification @stefanpenner 

I'll pass it on to the team

**Comment by icecoldmax** [comment]
@stefanpenner How about observing ANY property changes on things like `ObjectProxy`s? I've been doing `Ember.observer("foos.@each")` to catch any change (say the `checked` property for a checkbox). Will this still work with `Ember.observer("foos.[]")`?

**Comment by stefanpenner** [comment]
@icecoldmax I don't believe that is a feature. @each and [] are for arrays and array proxies. Not objects or object proxies.

I'm dubious that this works as you describe.

**Comment by icecoldmax** [comment]
@stefanpenner I think I described it wrong.

Let me try my checkbox example better:

```
proxiedFoos: Ember.computed("foos", function() {
  return this.get("foos")
    .map((foo) => {
      return Ember.ObjectProxy.create({
        content: foo,
        checked: false,
      });
  });
})
```

Let's say these are now iterated over as a list of checkboxes where checking it updates the `checked` to true. We want to do something when any of them are checked, so we've been successfully using `Ember.observer("foos.@each")`. Is this bad practice?

**Comment by stefanpenner** [comment]
> Ember.observer("proxiedFoos.@each").

since there is a map, proxiedFoos is always an array. 

* `proxiedFoos.@each` -> `proxiedFoos.[]`
* `proxiedFoos.@each.checked` is used for chaining, and thus stays the same

**Comment by aortbals** [comment]
I found that the easiest way to compare these various computed properties was live. I was slightly confused about the behavior. Here is a twiddle: http://ember-twiddle.com/02ddca9b6c89f2f8f92b.

TLDR, from my understanding:

- `array.[]` is for observing the array itself, such as adding or removing items and **is supported in 2.0**
- `array.@each.name` will recomputed if any of the items `name` prop is updated  and **is supported in 2.0**
- `array.@each` is deprecated, and **it does not observe all item attributes (I think this is what some users expect) in 1.12 or 1.13**. It works identically to `array.[]` in 1.13 from my testing
- You should use `array.[]` or `array.@each.somePropThatYouCareAbout` only

**Comment by icecoldmax** [comment]
@stefanpenner @aortbals Thanks for the clarification.

Is there actually a way to do what I'm thinking of? Say you have an array of objects, each with a ton of attributes and you want to observe when any of their properties change. Any way other than just observing/computing on `@each.propName`?

**Comment by stefanpenner** [comment]
> @each.propName?

no
